File: test-validation/java/xss/Safe.java
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.PrintWriter;
import java.io.IOException;

/**
 * 安全示例：XSS 防护
 *
 * 安全说明：
 * 对用户输入进行了 HTML 实体编码后再写入响应，
 * 确保特殊字符被转义，防止脚本注入。
 *
 * 预期检出：无（不应被检出为漏洞）
 */
public class Safe extends HttpServlet {

    /**
     * HTML 实体编码，防止 XSS
     */
    private String encodeHtml(String input) {
        if (input == null) {
            return "";
        }
        return input.replace("&", "&amp;")
                     .replace("<", "&lt;")
                     .replace(">", "&gt;")
                     .replace("\"", "&quot;")
                     .replace("'", "&#x27;");
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws IOException {
        response.setContentType("text/html;charset=UTF-8");
        PrintWriter out = response.getWriter();

        // 安全：对用户输入进行 HTML 编码后再输出
        String userName = request.getParameter("name");
        String safeName = encodeHtml(userName);
        out.println("<html><body>");
        out.println("<h1>Welcome, " + safeName + "</h1>");
        out.println("</body></html>");
        out.flush();
    }

    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws IOException {
        response.setContentType("text/html;charset=UTF-8");
        PrintWriter out = response.getWriter();

        // 安全：同样进行了 HTML 编码
        String comment = request.getParameter("comment");
        String safeComment = encodeHtml(comment);
        out.println("<html><body>");
        out.println("<div class='comment'>" + safeComment + "</div>");
        out.println("</body></html>");
        out.flush();
    }
}

File: test-validation/typescript/xss/safe.ts
/**
 * 安全示例：XSS 防护 - TypeScript
 *
 * 安全说明：
 * 使用 textContent 替代 innerHTML，或使用 DOMPurify 进行 HTML 消毒，
 * 确保用户输入不会被解释为可执行脚本。
 *
 * 预期检出：无（不应被检出为漏洞）
 */

/**
 * 安全：使用 textContent 替代 innerHTML
 */
function displayUserName(input: string): void {
  const container = document.getElementById("user-display");
  if (container) {
    // 安全：使用 textContent，浏览器不会解析 HTML 标签
    const heading = document.createElement("h1");
    heading.textContent = `Welcome, ${input}`;
    container.innerHTML = "";
    container.appendChild(heading);
  }
}

/**
 * 安全：使用 DOMPurify 消毒 HTML
 */
function renderComment(comment: string): void {
  // 安全：使用 DOMPurify 消毒后再赋值
  const container = document.getElementById("comments");
  if (container) {
    const div = document.createElement("div");
    div.className = "comment";
    div.textContent = comment;
    container.appendChild(div);
  }
}

/**
 * 安全：使用 DOM API 创建元素，避免 innerHTML
 */
function updateElement(userContent: string): void {
  const el = document.getElementById("content");
  if (el) {
    // 安全：使用 DOM API 创建元素
    const section = document.createElement("section");
    section.textContent = userContent;
    el.replaceWith(section);
  }
}

/**
 * 安全：手动 HTML 实体编码
 */
function encodeHtml(input: string): string {
  return input
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#x27;");
}

function displayWithEncoding(input: string): void {
  const container = document.getElementById("safe-display");
  if (container) {
    // 安全：先编码再赋值 innerHTML
    container.innerHTML = `<h1>Welcome, ${encodeHtml(input)}</h1>`;
  }
}
